pyomo/core/expr/numvalue.py

```diff
@@ -266,92 +247,139 @@ def potentially_variable(obj):
-                    return tmp
-            except:
-                pass
-
-            tmp = NumericConstant(obj)
-            update_KnownConstants(obj, tmp)
-            return tmp
+        #
+        # Because INT, FLOAT, and sometimes LONG hash the same, we
+        # index _KnownConstants with both the class and value.  This ensures
+        # consistent results.
+        #
+        val = _KnownConstants.get((obj.__class__, obj), None)
+        if val is not None:
+            return val 
+        #
+        # Create the numeric constant.  This really
+        # should be the only place in the code
+        # where these objects are constructed.
+        #
+        retval = NumericConstant(obj)
+        #
+        # Cache the numeric constants.  We used a bounded the cache size to avoid
+        # unexpectedly large lists of constants.  There are typically a small number
+        # of constants that need to be cached.
+        #
+        # NOTE:  A LFU policy might be more sensible here, but that
+        # requires a more complex cache.  It's not clear that that
+        # is worth the extra cost.
+        #
+        if len(_KnownConstants) < 1024:
+            _KnownConstants[obj.__class__,obj] = retval
+            return retval
+        #
+        return retval
+    #
+    # Ignore objects that are duck types to work with Pyomo expressions
+    #
     try:
-        return obj.as_numeric()
+        obj.is_expression_type()
+        return obj
     except AttributeError:
         pass
+    #
+    # Test if the object looks like a number.  If so, register that type with a
+    # warning.
+    #
     try:
         if obj.__class__ is (obj + 0).__class__:
+            #
             # obj may (or may not) be hashable, so we need this try
             # block so that things proceed normally for non-hashable
             # "numeric" types
+            #
+            retval = NumericConstant(obj)
             try:
-                if obj in KnownConstants:
-                    return KnownConstants[obj]
-                else:
-                    tmp = NumericConstant(obj)
-                    update_KnownConstants(obj, tmp)
-
-                    # If we get here, this is a reasonably well-behaving
-                    # numeric type: add it to the native numeric types
-                    # so that future lookups will be faster.
-                    native_numeric_types.add(obj.__class__)
-                    # native numeric types are also native types
-                    native_types.add(obj.__class__)
-
-                    return tmp
+                #
+                # Create the numeric constant and add to the 
+                # list of known constants.
+                #
+                # Note: we don't worry about the size of the
+                # cache here, since we need to confirm that the
+                # object is hashable.
+                #
+                _KnownConstants[obj.__class__,obj] = retval
```

Comment:
Editorial, but strongly recommended (and easy to implement):  In an attempt to reduce the amount of repeated code, instead of creating the constant here, what about just updating the sets, toss the warning, and then returning a recursive call?  That also has the benefit of avoiding needing to explain why we are ignoring the cache size here.

Reply:
This code is simpler because the code is repeated.  The recursive logic is less clear, and the duplicated code is < 10 lines.